repo/fsrepo/migrations/ipfsfetcher/ipfsfetcher.go

```diff
@@ -0,0 +1,288 @@
+	if err != nil {
+		return nil, err
+	}
+
+	fileNode, ok := nd.(files.File)
+	if !ok {
+		return nil, fmt.Errorf("%q is not a file", filePath)
+	}
+
+	if f.limit != 0 {
+		return migrations.NewLimitReadCloser(fileNode, f.limit), nil
+	}
+	return fileNode, nil
+}
+
+func (f *IpfsFetcher) Close() error {
+	f.closeOnce.Do(func() {
+		if f.ipfsStopFunc != nil {
+			// Tell ipfs node to stop and wait for it to stop
+			f.ipfsStopFunc()
+		}
+
+		if f.ipfsTmpDir != "" {
+			// Remove the temp ipfs dir
+			f.closeErr = os.RemoveAll(f.ipfsTmpDir)
+		}
+	})
+	return f.closeErr
+}
+
+func initTempNode(ctx context.Context) (string, error) {
+	err := setupPlugins()
+	if err != nil {
+		return "", err
+	}
+
+	identity, err := config.CreateIdentity(ioutil.Discard, []options.KeyGenerateOption{
+		options.Key.Type(options.Ed25519Key),
+	})
+	if err != nil {
+		return "", err
+	}
+	cfg, err := config.InitWithIdentity(identity)
+	if err != nil {
+		return "", err
+	}
+
+	// create temporary ipfs directory
+	dir, err := ioutil.TempDir("", "ipfs-temp")
+	if err != nil {
+		return "", fmt.Errorf("failed to get temp dir: %s", err)
+	}
+
+	// configure the temporary node
+	cfg.Routing.Type = "dhtclient"
+
+	// Disable listening for inbound connections
+	cfg.Addresses.Gateway = []string{}
+	cfg.Addresses.API = []string{}
+	cfg.Addresses.Swarm = []string{}
+
+	err = fsrepo.Init(dir, cfg)
+	if err != nil {
+		os.RemoveAll(dir)
+		return "", fmt.Errorf("failed to init ephemeral node: %s", err)
+	}
+
+	return dir, nil
+}
+
+func startTempNode(repoDir string, peers []peer.AddrInfo) (iface.CoreAPI, func(), error) {
+	// Open the repo
+	r, err := fsrepo.Open(repoDir)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create a new lifetime context that is used to stop the temp ipfs node
+	ctxIpfsLife, cancel := context.WithCancel(context.Background())
+
+	// Construct the node
+	node, err := core.NewNode(ctxIpfsLife, &core.BuildCfg{
+		Online:  true,
+		Routing: libp2p.DHTClientOption,
+		Repo:    r,
+	})
+	if err != nil {
+		cancel()
+		r.Close()
+		return nil, nil, err
+	}
+
+	ifaceCore, err := coreapi.NewCoreAPI(node)
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+
+	stopFunc := func() {
+		// Tell ipfs to stop
+		cancel()
+		// Wait until ipfs is stopped
+		<-node.Context().Done()
+	}
+
+	// Parse peer addresses and asynchronously connect to peers
+	if len(peers) != 0 {
+		connectPeers(ctxIpfsLife, ifaceCore, peers)
+	}
+
+	return ifaceCore, stopFunc, nil
+}
+
+func parsePath(fetchPath string) (ipath.Path, error) {
+	ipfsPath := ipath.New(fetchPath)
+	if ipfsPath.IsValid() == nil {
+		return ipfsPath, nil
+	}
+
+	u, err := url.Parse(fetchPath)
+	if err != nil {
+		return nil, fmt.Errorf("%q could not be parsed: %s", fetchPath, err)
+	}
+
+	switch proto := u.Scheme; proto {
+	case "ipfs", "ipld", "ipns":
+		ipfsPath = ipath.New(path.Join("/", proto, u.Host, u.Path))
+	case "http", "https":
+		ipfsPath = ipath.New(u.Path)
```

Comment:
Why would we get an HTTP(S) URL here? If we can't then let's remove this code, if we can then can let's discuss our options since this seems it will get us into trouble (e.g. what about subdomain gateways, maybe path gateways can start with some prefix like foo.tld/custom/ipfs/Qmxyz).

Reply:
Removed.  If someone were to try to use the IpfsFetcher with an HTTP path, then returning an is acceptable, and they really mean to use the HttpFetcher.